src/reducers/columnStateHelper.js

```diff
@@ -47,79 +42,180 @@ function initialize(state, props, oldProps) {
+    props.scrollToColumn !== null &&
+    (!oldProps || props.scrollToColumn !== oldProps.scrollToColumn)
+  ) {
+    return scrollToColumn(state, props.scrollToColumn);
   }
 
+  // if scrollLeft changed
+  if (
+    props.scrollLeft !== undefined &&
+    props.scrollLeft !== null &&
+    (!oldProps || props.scrollLeft !== oldProps.scrollLeft)
+  ) {
+    return scrollToPos(state, props.scrollLeft);
+  }
+
+  // no change in column anchor
+  return {
+    firstIndex: state.firstColumnIndex,
+    firstOffset: state.firstColumnOffset,
+    lastIndex: undefined,
+    changed: false,
+  };
+}
+
+/**
+ * @param {!Object} state
+ * @param {number} scrollToColumn
+ * @return {{
+ *   firstIndex: number,
+ *   firstOffset: number,
+ *   lastIndex: number,
+ *   changed: boolean,
+ * }}
+ */
+function scrollToColumn(state, scrollToColumn) {
   const {
     availableScrollWidth,
     fixedColumns,
     scrollableColumns,
   } = columnWidths(state);
-  const fixedColumnsCount = fixedColumns.length;
-  const scrollableColumnsCount = scrollableColumns.length;
-
-  const noScrollableColumns = scrollableColumnsCount === 0;
-  const scrollToUnchanged = scrollToColumn === oldScrollToColumn;
-  const selectedColumnFixed = scrollToColumn < fixedColumnsCount;
-  const selectedColumnFixedRight = scrollToColumn >= fixedColumnsCount + scrollableColumnsCount;
-  if (scrollToUnchanged || selectedColumnFixed || selectedColumnFixedRight || noScrollableColumns) {
-    return scrollX;
+
+  const { scrollX, columnOffsetIntervalTree } = state;
+
+  const selectedColumnFixed = scrollToColumn < fixedColumns.length;
+  const selectedColumnFixedRight = scrollToColumn >= fixedColumns.length + scrollableColumns.length;
+
+  // if target column is fixed, then don't do anything
+  if (selectedColumnFixed || selectedColumnFixedRight) {
+    return {
+      firstIndex: state.firstColumnIndex,
+      firstOffset: state.firstColumnOffset,
+      lastIndex: undefined,
+      changed: false,
+    }
   }
 
-  // Convert column index (0 indexed) to scrollable index (0 indexed)
-  // and clamp to max scrollable index
-  const clampedColumnIndex = Math.min(scrollToColumn - fixedColumnsCount,
-    scrollableColumns.length - 1);
+  // convert to scrollable column index where 0 denotes first scrollable column
+  const clampedColumnIndex = Math.min(scrollToColumn - fixedColumns.length, scrollableColumns.length - 1);
+  const columnBeginX = columnOffsetIntervalTree.sumUntil(clampedColumnIndex);
+  const columnEndX = columnBeginX + updateColumnWidth(state, clampedColumnIndex);
 
-  // Compute the width of all columns to the left of the column
-  let previousWidth = 0;
-  for (let columnIdx = 0; columnIdx < clampedColumnIndex; ++columnIdx) {
-    previousWidth += scrollableColumns[columnIdx].width;
+  let firstIndex = clampedColumnIndex;
+  let lastIndex = undefined;
+  if (columnBeginX < scrollX) {
+    // If column starts before the viewport, set as the first column in the viewport
```

Comment:
Why not initialize them inside?

Reply:
> Why not initialize them inside?

hmm, I'll take a look at this, but we'll probably have to use return at 3 different points, and with 4 properties to return it might be too much

Reply:
Having many returns won't be too bad since it doesn't look like they share logic.  We could even use the shortcutting style to make it more readable:
```
if (A) {
  return blah;
}
if (B) {
  return foo;
}
return bar;
```